### ISSUE TITLE
[JENKINS-12092] Allow a job to completely block a category

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/throttle-concurrent-builds-plugin#345]
<p>It'd be good if a job could be set up to completely block an entire category. Use-case for this would be having a "writer" job that e.g. updates a server, and having a set of "reader" jobs that contact that server. The "writer" job would completely block the "use-server" category from running <b>any</b> concurrent builds, and the "reader" jobs would just use 1 slot of the allocated slots for that category.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jorgenpt.jpg" title="jorgenpt's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jorgenpt">jorgenpt</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12092" target="_blank">Allow a job to completely block a category</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: throttle-concurrent-builds-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 3
<li><b>watchers</b>: 3
<li><b>imported</b>: 20251218-075729
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>It'd be good if a job could be set up to completely block an entire category. Use-case for this would be having a "writer" job that e.g. updates a server, and having a set of "reader" jobs that contact that server. The "writer" job would completely block the "use-server" category from running <b>any</b> concurrent builds, and the "reader" jobs would just use 1 slot of the allocated slots for that category.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>fluffy88:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12092?focusedId=219349&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-219349">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I would like to see this feature also.</p>

<p>We have,</p>
<ul>
	<li>Multiple jobs setup in a pipeline, some of which run on labelled slaves</li>
	<li>A job that refreshes the labelled slave
(i.e. deletes and creates a new one)</li>
</ul>


<p>The refresh job and the pipeline jobs cannot run at the same time or disaster will ensue.</p>

<p>So some way to throttle the refresh job from executing while the pipeline jobs are running would be very useful.</p>
</pre>
</details>

</ul>
</details>

<p>I would like to see this feature also.</p>

<p>We have,</p>
<ul>
	<li>Multiple jobs setup in a pipeline, some of which run on labelled slaves</li>
	<li>A job that refreshes the labelled slave<br/>
(i.e. deletes and creates a new one)</li>
</ul>


<p>The refresh job and the pipeline jobs cannot run at the same time or disaster will ensue.</p>

<p>So some way to throttle the refresh job from executing while the pipeline jobs are running would be very useful.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ki82.png" title="ki82's avatar" /> ki82:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12092?focusedId=219669&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-219669">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I think the ideal way to solve this is to have a multi-project throttle category that you assign to all your jobs.
The multi-project throttle category should be able to have high limitations or no limitations at all.</p>

<p>If you then create:
jobA and jobB with the limitations 0/0 they should be able to run concurrently up to the multi-project throttle category limitations.
jobC with the limitations 1 per node, jobC should only run on a node if no other job with the throttle category runs on the node.
jobD with the limitations 1 in total, jobD should only run if no other jobs of this category is running.</p>

<p>The UI to support this is already there, I have tried to set this up since I thought that it should work before I saw this issue <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>I think the ideal way to solve this is to have a multi-project throttle category that you assign to all your jobs.<br/>
The multi-project throttle category should be able to have high limitations or no limitations at all.</p>

<p>If you then create:<br/>
jobA and jobB with the limitations 0/0 they should be able to run concurrently up to the multi-project throttle category limitations.<br/>
jobC with the limitations 1 per node, jobC should only run on a node if no other job with the throttle category runs on the node.<br/>
jobD with the limitations 1 in total, jobD should only run if no other jobs of this category is running.</p>

<p>The UI to support this is already there, I have tried to set this up since I thought that it should work before I saw this issue <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ki82.png" title="ki82's avatar" /> ki82:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12092?focusedId=219670&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-219670">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>300$ is up for grabs for solving this issue at: <a href="https://freedomsponsors.org/issue/618/allow-a-job-to-completely-block-a-category?alert=SPONSOR#" class="external-link" target="_blank" rel="nofollow noopener">https://freedomsponsors.org/issue/618/allow-a-job-to-completely-block-a-category?alert=SPONSOR#</a></p>
</pre>
</details>

</ul>
</details>

<p>300$ is up for grabs for solving this issue at: <a href="https://freedomsponsors.org/issue/618/allow-a-job-to-completely-block-a-category?alert=SPONSOR#" class="external-link" target="_blank" rel="nofollow noopener">https://freedomsponsors.org/issue/618/allow-a-job-to-completely-block-a-category?alert=SPONSOR#</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>fluffy88:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12092?focusedId=220102&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-220102">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Christian,</p>

<p>I don't believe the UI that's presented in the jobs configuration is there for this purpose. As far as I can see, if you select <em>Throttle this project as part of one or more categories</em> the 2 <em>Maximum Concurrent Builds</em> textboxes are ignored. Those textboxes are only used when the <em>Throttle this project alone</em> radio is selected.
At least, from my limited understanding of the plugin that's how it works.</p>

<p>I've been spending a bit of time on this issue.
The solution I have implemented so far is, </p>
<ul>
	<li>I added a new checkbox that belongs to the <em>Throttle this project as part of one or more categories</em> radio</li>
	<li>The new checkbox <em>Let this Job be blocked by selected Categories (it's a working title)</em> will block a job if any builds from any of the selected categories are queued or executing</li>
</ul>


<p>This in effect lets you select 1 or more jobs that will not execute while jobs from it's categories are executing.
What do you think of this solution?
I'd like to get some community feedback before I make a pull request.
Thanks!</p>

<p>P.S:
I was also considering fixing the UI so the "child" options only show up when you select it's parent radio button. But that's part of a different issue.</p>
</pre>
</details>

</ul>
</details>

<p>Hi Christian,</p>

<p>I don't believe the UI that's presented in the jobs configuration is there for this purpose. As far as I can see, if you select <em>Throttle this project as part of one or more categories</em> the 2 <em>Maximum Concurrent Builds</em> textboxes are ignored. Those textboxes are only used when the <em>Throttle this project alone</em> radio is selected.<br/>
At least, from my limited understanding of the plugin that's how it works.</p>

<p>I've been spending a bit of time on this issue.<br/>
The solution I have implemented so far is, </p>
<ul>
	<li>I added a new checkbox that belongs to the <em>Throttle this project as part of one or more categories</em> radio</li>
	<li>The new checkbox <em>Let this Job be blocked by selected Categories (it's a working title)</em> will block a job if any builds from any of the selected categories are queued or executing</li>
</ul>


<p>This in effect lets you select 1 or more jobs that will not execute while jobs from it's categories are executing.<br/>
What do you think of this solution?<br/>
I'd like to get some community feedback before I make a pull request.<br/>
Thanks!</p>

<p>P.S:<br/>
I was also considering fixing the UI so the "child" options only show up when you select it's parent radio button. But that's part of a different issue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ki82.png" title="ki82's avatar" /> ki82:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12092?focusedId=220217&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-220217">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Seán,</p>

<p>Thanks for looking into this.</p>

<p>What I am missing from your solution is the ability to throttle to 1 instance per-node not only on a global level.</p>

<p>I think that it would be more dynamic if the original UI boxes could be used and that one could use 1 per node or 1 in total or any other combination of node/total that is lower than the category limit.
Since we don't need it currently it's not needed but nice-to-have.</p>
</pre>
</details>

</ul>
</details>

<p>Hi Seán,</p>

<p>Thanks for looking into this.</p>

<p>What I am missing from your solution is the ability to throttle to 1 instance per-node not only on a global level.</p>

<p>I think that it would be more dynamic if the original UI boxes could be used and that one could use 1 per node or 1 in total or any other combination of node/total that is lower than the category limit.<br/>
Since we don't need it currently it's not needed but nice-to-have.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>fluffy88:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12092?focusedId=220227&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-220227">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for the feedback Christian!</p>

<p>I don't know if using the threshold inputs on the Job configuration page is the right place to do this. </p>

<p>But I think we could implement this same support by moving the "blocking" part from the Job configuration into the Global configuration where we define the Categories. If we allowed category B to be blocked by category A.
This way we could allow category A to completely block all jobs in category B from executing until all queued jobs from category A have completed.</p>

<p>This would allow both categories to have their own separate throttle limits which would not interfere with each other.</p>
</pre>
</details>

</ul>
</details>

<p>Thanks for the feedback Christian!</p>

<p>I don't know if using the threshold inputs on the Job configuration page is the right place to do this. </p>

<p>But I think we could implement this same support by moving the "blocking" part from the Job configuration into the Global configuration where we define the Categories. If we allowed category B to be blocked by category A.<br/>
This way we could allow category A to completely block all jobs in category B from executing until all queued jobs from category A have completed.</p>

<p>This would allow both categories to have their own separate throttle limits which would not interfere with each other.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ki82.png" title="ki82's avatar" /> ki82:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12092?focusedId=220372&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-220372">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>That sounds like an good idea Seán.</p>
</pre>
</details>

</ul>
</details>

<p>That sounds like an good idea Seán.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>fluffy88:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12092?focusedId=220648&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-220648">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Pull request made.
<a href="https://github.com/jenkinsci/throttle-concurrent-builds-plugin/pull/25" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/throttle-concurrent-builds-plugin/pull/25</a></p>
</pre>
</details>

</ul>
</details>

<p>Pull request made.<br/>
<a href="https://github.com/jenkinsci/throttle-concurrent-builds-plugin/pull/25" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/throttle-concurrent-builds-plugin/pull/25</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ki82.png" title="ki82's avatar" /> ki82:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-12092?focusedId=223182&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-223182">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Sean, are you working on resolving the merge conflicts to get it into the repo?</p>
</pre>
</details>

</ul>
</details>

<p>Sean, are you working on resolving the merge conflicts to get it into the repo?</p>
